pkg/api/write.go

```diff
@@ -15,188 +14,187 @@ import (
+		m.SentSamples.Add(float64(numSamples))
+		m.SentBatchDuration.Observe(duration)
 
-		metrics.WriteThroughput.SetCurrent(getCounterValue(metrics.SentSamples))
+		m.WriteThroughput.SetCurrent(getCounterValue(m.SentSamples))
 
 		select {
-		case d := <-metrics.WriteThroughput.Values:
+		case d := <-m.WriteThroughput.Values:
 			log.Info("msg", "Samples write throughput", "samples/sec", d)
 		default:
 		}
 
-	})
+		return true
+	}
 }
 
-func loadWriteRequest(r *http.Request) (*prompb.WriteRequest, error, string) {
-	var req *prompb.WriteRequest
-
-	switch r.Header.Get("Content-Type") {
-	case "application/x-protobuf":
-		reqBuf, err, msg := decodeSnappyBody(r.Body)
-		if err != nil {
-			return nil, err, msg
-		}
-
-		req = ingestor.NewWriteRequest()
-		if err := proto.Unmarshal(reqBuf, req); err != nil {
-			ingestor.FinishWriteRequest(req)
-			return nil, err, "Protobuf unmarshal error"
-		}
-	case "application/json":
-		var (
-			i    jsonPayload
-			body io.Reader = r.Body
-			err  error
-			msg  string
-		)
-
-		snappyEncoding := strings.Contains(r.Header.Get("Content-Encoding"), "snappy")
-
-		if snappyEncoding {
-			body, err, msg = decodeSnappy(r.Body)
+type writeStage func(http.ResponseWriter, *http.Request, *Metrics) bool
 
-			if err != nil {
-				return nil, err, msg
-			}
-		}
-
-		dec := json.NewDecoder(body)
-		req = ingestor.NewWriteRequest()
-
-		for {
-			if err := dec.Decode(&i); err == io.EOF {
-				break
-			} else if err != nil {
-				ingestor.FinishWriteRequest(req)
-				return nil, err, "JSON decode error"
-			}
-
-			req = appendJSONPayload(req, i)
-		}
-	default:
-		// Cannot get here because of header validation.
-		return nil, fmt.Errorf("unsupported data format (not protobuf or json)"), "Write header validation error"
-	}
-
-	return req, nil, ""
+type writeHandler struct {
+	stages  []writeStage
+	metrics *Metrics
 }
 
-func isWriter(elector *util.Elector) (bool, error) {
-	if elector != nil {
-		shouldWrite, err := elector.IsLeader()
-		return shouldWrite, err
+func (wh *writeHandler) addStages(stages ...writeStage) {
+	for _, stage := range stages {
+		if stage == nil {
+			continue
+		}
+		wh.stages = append(wh.stages, stage)
 	}
-	return true, nil
 }
 
-func decodeSnappy(r io.Reader) (io.Reader, error, string) {
-	buf := make([]byte, 10)
-	_, err := r.Read(buf)
-	if err != nil {
-		return nil, err, "snappy payload too short or corrupt"
-	}
-
-	//Setup multi-reader so we can read the complete contents.
-	mr := io.MultiReader(bytes.NewBuffer(buf), r)
-
-	if detectSnappyStreamFormat(buf) {
-		return snappy.NewReader(mr), nil, ""
-	}
-
-	// Snappy block format.
-	reqBuf, err, msg := decodeSnappyBody(mr)
-	if err != nil {
-		return nil, err, msg
-	}
-
-	return bytes.NewBuffer(reqBuf), nil, ""
+func (wh *writeHandler) handler() http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		for _, stage := range wh.stages {
+			if !stage(w, r, wh.metrics) {
+				return
```

Comment:
I think we should also check for errors along with bool, and then report the error with the stage information. This will help understand the error better.

Reply:
I don't understand this comment. The stages handle errors themselves, there is no general way to handle them. They are the top level code which needs to log and handle based on the circumstance.